README.md
## DEPRECATED

The project got superseded by the [CodeImageGenerator](https://github.com/peitek/CodeImageGenerator) project, which creates generally usable images instead of Presentation-specific output.

# PresentationCodeFormatter

PresentationCodeFormatter is a Python script, which converts Java files/functions to a text file, which is readable by [Presentation](http://www.neurobs.com/).
This includes a rough custom syntax highlighting with [Pygments](http://pygments.org/).

I personally use this to convert short code snippets for empirical research on program comprehension (see [Brains on Code](https://github.com/brains-on-code)).
This means the script is personalized to my specific usage and probably not directly useful to you without changes.

Disclaimer: I am not a Python programmer. I'm always grateful for feedback on how to improve my code.

## Features

The main feature is to read all files from a directory and attempting to add syntax highlighting compatible with the syntax of Presentation to the output file(s).

Implemented so far:

* Reads all files from a directory, and parses one at a time
* Extracts just the function code from the file (and ignores the class, comments, etc around it)
* Uses Pygments to create syntax highlighting for each function
* Parses Pygments syntax highlighting to a Presentation-compatible format (still may be incomplete)
* Wraps code into a Presentation variable for seamless integration
* Output in separate files, or one large file


## Roadmap

Things I still want to add:

* Command line support
* Better configuration (in particular the output color scheme)
* Allow multiple functions per file
* Allow functions besides `public int`, `public String`, `public boolean`
* Only read Java files
* Support for more languages than just Java


## Setup ##

The project should run in any Python environment. It was developed and tested with the [PyCharms IDE](https://www.jetbrains.com/pycharm/).


# License #

```
MIT License

Copyright (c) 2017 Norman Peitek
```

CHANGELOG.md
# Changelog #

## 0.2.1 (2018-04-11)

* Move image generator to its [own repository](https://github.com/peitek/CodeImageGenerator)
* Deprecate this repository

## 0.2.0 (2017-11-17)

* Add experimental option to export code with syntax highlighting as PNG image

## 0.1.1 (2017-06-03)

* Put exported snippets into executable script variables
* Add export of pcl file for all snippets

## 0.1.0 (2017-06-02)

- Created first version

export_as_image.py
import PIL
from PIL import Image, ImageDraw, ImageFont

import html

import os
from os.path import join

import pygments
import pygments.lexers as lexers
import pygments.formatters as formatters
import pygments.styles as styles

FONT_COLOR_INSTRUCTION = (200, 200, 200)


def main():
    code = """
public boolean containsSubstring(String word, String substring) {
    for (int i = 0; i < word.length(); i++) {
        for (int j = 0; j < substring.length(); j++) {
            if (i + j >= word.length())
                break;
            if (word.charAt(i + j) != substring.charAt(j)) {
                break;
            } else {
                if (j == substring.length() - 1) {
                    return true;
                }
            }
        }
    }

    return false;
}
"""
    getImage("test123", code)


def getImage(function_name, code, code_task):
    print("creating image for " + function_name)
    fullSizeX = 1280 #1920
    fullSizeY = 1024 #1080
    image = Image.new("RGBA", (fullSizeX, fullSizeY), (0,0,0))
    draw = ImageDraw.Draw(image)

    inconsolata = ImageFont.truetype("resources/Inconsolata-Regular.ttf", 32)

    # draw instructions for code task
    (instruction_width, instruction_height) = ImageDraw.ImageDraw(image).textsize(text=code_task, font=inconsolata)
    draw.text((fullSizeX/2 - instruction_width/2, (fullSizeY / 20)), code_task, FONT_COLOR_INSTRUCTION, font=inconsolata)

    # draw code
    code_in_html = create_syntax_highlighting_html(code)
    code_in_html = html.unescape(code_in_html)
    code_in_html = code_in_html.replace('\t', '    ')
    code_in_html = code_in_html.replace("<span></span>", "")

    (allTextSizeX, allTextSizeY) = ImageDraw.ImageDraw(image).multiline_textsize(text=code, font=inconsolata)
    allTextSizeY *= 1.4 #for additional vertical padding, otherwise the text is too hard to read

    xPosStart = (fullSizeX/2) - (allTextSizeX/2)
    yPos = (fullSizeY/2) - (allTextSizeY/2)
    (nil, verticalPadding) = ImageDraw.ImageDraw(image).textsize(text="blubb", font=inconsolata)

    code_in_html_lines = code_in_html.split("\n")

    for code_line in code_in_html_lines:
        if code_line == "</pre></div>":
            break

        # remove four first characters from each line to offset the incorrect indentation
        if code_line.startswith('    '):
            code_line = code_line[4:]

        code_elements = code_line.split("<span")

        xPos = xPosStart
        yPos += (verticalPadding * 1.4)

        for element in code_elements:
            element = element.replace("</span>", "")
            endPos = element.find('>')

            span_class = ''

            if endPos != -1:
                span_class = element[8:endPos-1]
                element = element[endPos+1:]

            draw.text((xPos, yPos), element, get_color_for_span_class(span_class), font=inconsolata)
            (sizeX, sizeY) = ImageDraw.ImageDraw(image).textsize(text=element, font=inconsolata)
            xPos += sizeX

    write_image_to_file(function_name, image)


def write_image_to_file(file_name, image):
    subdirectory = 'output_images'
    try:
        os.mkdir(subdirectory)
    except Exception:
        pass

    #image.thumbnail((1280, 1024), PIL.Image.ANTIALIAS)
    #image.save(join(subdirectory, file_name + '_resized.png'), "PNG")
    image.save(join(subdirectory, file_name + '.png'), "PNG")


def get_color_for_span_class(span_class):
    return {
        'kc': (0,128,0),
        'kd': (0,0,255),
        'k': (0,128,64),
        'kt': (67,168,237),
        'mi': (85,26,139),
        'nf': (102, 255, 255),
        'na': (125,144,41),
        'String': (125,144,41),
        'o': (102,102,102),
        's': (0,128,0),
        'sc': (0,128,0),
        'n': (255,255,255),
        'err': (255,255,255),
    }.get(span_class, (255,255,255))


def create_syntax_highlighting_html(code_function_string):
    # Create Pygments formatter (http://pygments.org/)
    formatter = formatters.get_formatter_by_name("html")
    formatter.full = True
    formatter.style = styles.get_style_by_name("manni")
    lexer = lexers.get_lexer_by_name('Java')
    # Use pygments to create code with syntax highlighting in HTML
    code_in_html = pygments.highlight(code_function_string, lexer, formatter)
    # Convert syntax highlighting from HTML to the Presentation-format
    code_in_presentation = code_in_html
    pos = code_in_html.find("<pre>")
    code_in_presentation = code_in_presentation.replace(code_in_presentation[:pos + 5], '')
    return code_in_presentation


main.py
import os
from os.path import isfile, join

import pygments
import pygments.lexers as lexers
import pygments.formatters as formatters
import pygments.styles as styles


def main():
    # TODO let this program be executable from the command line
    # TODO make config accessible from outside
    # config
    file_directory = "C:/Users/npeitek/Documents/fmri-td/CodeSnippets/src/com/fmri/topdown/original/number"
    limit_to_files_with_condition = False
    output_separated = False
    output_single = {}
    output_single_name = 'AllFunctionsWords'
    snippet_time = '$TopDownTime'

    # read all files from a directory and loop through them
    only_files = [f for f in os.listdir(file_directory) if isfile(join(file_directory, f))]

    for file_name in only_files:
        if limit_to_files_with_condition:
            # only select files with a condition in their name
            conditions = ['BU', 'LOBO', 'LOBS']
            if any(x in file_name for x in conditions):
                convert_file(file_directory, file_name, output_single, output_separated, snippet_time)
        else:
            convert_file(file_directory, file_name, output_single, output_separated, snippet_time)

    # if configured, write large output file with everything put together
    if not output_separated:
        all_functions = ''
        function_calls = ''

        for key, value in output_single.items():
            all_functions += value + '\n\n'
            function_calls += key + '.present();\n'

        write_presentation_string_to_file(all_functions, output_single_name)
        write_presentation_string_to_file(function_calls, output_single_name + "_pclfile")


def convert_file(file_directory, file_name, output, output_separated, snippet_time):
    with open(join(file_directory, file_name)) as text_file:
        code_file = text_file.read()

        # extract Java function from code file
        try:
            code_function_start, code_function_string, code_task = extract_function_from_file(code_file)
        except Exception:
            return

        # get name of the function to specify the Presentation code block
        # TODO figure out unscrambled function name
        function_name_string = code_function_string[len(code_function_start):]
        function_name_position_end = function_name_string.find("(")
        function_name = function_name_string[:function_name_position_end]

        # remove conditions (LDBO, ..) from function name
        # TODO limit it to the function name to prevent bugs
        code_function_string = code_function_string \
            .replace('TD_B', '') \
            .replace('TD_N', '') \
            .replace('TD_U', '') \
            .replace('LDBO', '') \
            .replace('LDBS', '') \
            .replace('LOBO', '') \
            .replace('LOBS', '') \
            .replace('BU', '') \
            .replace('TD', '') \
            .replace('SY', '')

        code_in_html = create_syntax_highlighting_html(code_function_string)

        code_in_presentation = convert_syntax_highlighting_to_presentation(code_in_html)

        # Put the code with syntax highlighting in Presentation's variable framework
        full_presentation_string = "#" + function_name + """ 
picture {    
    text { 
        formatted_text = true;
        caption = \"<font color='200,200,200'>          equivalent?</font>
                        
                        
""" + code_in_presentation + """";}; x = $xf; y = $yf;}
code_""" + function_name + "; \n\n" + \
"""
trial {
    trial_duration = '""" + snippet_time + """';
    picture code_""" + function_name + """ ; 
    time = 0 ; 
    duration = '$TopDownTime'; 
    code = \"code_""" + function_name + """\"; 
} """ + function_name + """;"""

        # TODO move all files into an array

        # TODO write entire array into an output file
        if output_separated:
            write_presentation_string_to_file(full_presentation_string, function_name)
        else:
            output[function_name] = full_presentation_string


def extract_function_from_file(code_file):
    # TODO support multiple functions for each file
    code_function_start_lookups = ['public int ', 'public String ', 'public Integer ', 'public float ', 'public boolean ', 'public double[] ', 'public int[] ']

    code_function_start = next((x for x in code_function_start_lookups if x in code_file), False)

    if not code_function_start:
        raise Exception('no function found')

    code_task_position = code_file.find("[TASK]")

    if code_task_position != -1:
        code_task_string = code_file[code_task_position + 7:]
        code_task_end = code_task_string.find("\n")
        code_task_string = code_task_string[:code_task_end]
    else:
        code_task_string = ""

    code_function_position = code_file.find(code_function_start)
    code_function_string = code_file[code_function_position:]
    number_of_open_curly_brackets = 0
    code_function_end = -1
    for i, c in enumerate(code_function_string):
        if c == "{":
            number_of_open_curly_brackets += 1

        if c == "}":
            number_of_open_curly_brackets -= 1
            if (number_of_open_curly_brackets <= 0):
                code_function_end = i
                break
    code_function_string = code_function_string[:code_function_end + 1]
    return code_function_start, code_function_string, code_task_string


def create_syntax_highlighting_html(code_function_string):
    # Create Pygments formatter (http://pygments.org/)
    formatter = formatters.get_formatter_by_name("html")
    formatter.full = True
    formatter.style = styles.get_style_by_name("manni")
    lexer = lexers.get_lexer_by_name('Java')

    # Use pygments to create code with syntax highlighting in HTML
    code_in_html = pygments.highlight(code_function_string, lexer, formatter)

    # Convert syntax highlighting from HTML to the Presentation-format
    code_in_presentation = code_in_html
    pos = code_in_html.find("<pre>")
    code_in_presentation = code_in_presentation.replace(code_in_presentation[:pos + 5], '')
    return code_in_presentation


def convert_syntax_highlighting_to_presentation(code_in_presentation):
    # TODO handle all other <span> classes (ignore them?)
    # TODO move colors into separate configuration
    code_in_presentation = code_in_presentation.replace("<span></span>", "")
    code_in_presentation = code_in_presentation.replace('<span class="k">', "<font color='0,128,0'>")  # "for"
    code_in_presentation = code_in_presentation.replace('<span class="kc">', "<font color='0,128,0'>")
    code_in_presentation = code_in_presentation.replace('<span class="kd">',
                                                        "<font color='0,0,255'>")  # public, private...
    code_in_presentation = code_in_presentation.replace('<span class="kt">',
                                                        "<font color='67,168,237'>")  # function type (void, ..)
    code_in_presentation = code_in_presentation.replace('<span class="mi">',
                                                        "<font color='85,26,139'>")  # variable declaration value like = 1
    code_in_presentation = code_in_presentation.replace('<span class="n">', "<font color='255,255,255'>")  # neutral
    code_in_presentation = code_in_presentation.replace('<span class="err">', "<font color='255,255,255'>")  # neutral
    code_in_presentation = code_in_presentation.replace('<span class="nf">',
                                                        "<font color='102, 255, 255'>")  # function name
    code_in_presentation = code_in_presentation.replace('<span class="na">',
                                                        "<font color='125,144,41'>")  # variable type
    code_in_presentation = code_in_presentation.replace('String',
                                                        "<font color='125,144,41'>String</font>")  # variable type
    code_in_presentation = code_in_presentation.replace('<span class="o">',
                                                        "<font color='102,102,102'>")  # operator, brackets
    code_in_presentation = code_in_presentation.replace('<span class="s">', "<font color='0,128,0'>")
    code_in_presentation = code_in_presentation.replace('<span class="sc">', "<font color='0,128,0'>")  # literal: char
    code_in_presentation = code_in_presentation.replace('&quot;', '\\"')
    code_in_presentation = code_in_presentation.replace('&#39;', '\'')
    code_in_presentation = code_in_presentation.replace('</span>', '</font>')
    code_in_presentation = code_in_presentation[:-29]
    return code_in_presentation


def write_presentation_string_to_file(full_presentation_string, function_name):
    subdirectory = 'output'
    try:
        os.mkdir(subdirectory)
    except Exception:
        pass
    with open(join(subdirectory, 'Output_' + function_name + '.txt'), 'w') as output_file:
        print(full_presentation_string, file=output_file)


# Just call main function at the moment
main()
